osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs

```diff
@@ -144,54 +201,222 @@ protected override void LoadComplete()
             {
                 if (newVolume == null)
                     return;
 
-                beatmap.BeginChange();
+                update((_, samples) =>
+                {
+                    for (int i = 0; i < samples.Count; i++)
+                    {
+                        samples[i] = samples[i].With(newVolume: newVolume.Value);
+                    }
+                });
+            }
+
+            #region hitsound toggles
+
+            private readonly Dictionary<string, Bindable<TernaryState>> selectionSampleStates = new Dictionary<string, Bindable<TernaryState>>();
 
-                foreach (var h in objects)
+            private void createStateBindables()
+            {
+                foreach (string sampleName in HitSampleInfo.AllAdditions)
                 {
-                    for (int i = 0; i < h.Samples.Count; i++)
+                    var bindable = new Bindable<TernaryState>
+                    {
+                        Description = sampleName.Replace("hit", string.Empty).Titleize()
+                    };
+
+                    bindable.ValueChanged += state =>
                     {
-                        h.Samples[i] = h.Samples[i].With(newVolume: newVolume.Value);
+                        switch (state.NewValue)
+                        {
+                            case TernaryState.False:
+                                removeHitSample(sampleName);
+                                break;
+
+                            case TernaryState.True:
+                                addHitSample(sampleName);
+                                break;
+                        }
+                    };
+
+                    selectionSampleStates[sampleName] = bindable;
+                }
+            }
+
+            private void updateTernaryStates()
+            {
+                foreach ((string sampleName, var bindable) in selectionSampleStates)
+                {
+                    bindable.Value = SelectionHandler<HitObject>.GetStateFromSelection(relevantObjects, h => GetSamples(h).Any(s => s.Name == sampleName));
+                }
+            }
+
+            private IEnumerable<TernaryButton> createTernaryButtons()
+            {
+                foreach ((string sampleName, var bindable) in selectionSampleStates)
+                    yield return new TernaryButton(bindable, string.Empty, () => ComposeBlueprintContainer.GetIconForSample(sampleName));
+            }
+
+            private void addHitSample(string sampleName)
+            {
+                if (string.IsNullOrEmpty(sampleName))
+                    return;
+
+                update((h, samples) =>
+                {
+                    // Make sure there isn't already an existing sample
+                    if (samples.Any(s => s.Name == sampleName))
+                        return;
+
+                    // First try inheriting the sample info from the node samples instead of the samples of the hitobject
```

Comment:
I had a lot of trouble understanding how this changed anything until I realised that the `samples` list here is actually augmented from the `HitObject`'s samples.